subsys/bluetooth/mesh/gen_plvl_srv.c

```diff
@@ -0,0 +1,570 @@
+	{ GEN_PLVL_OP_DEFAULT_GET, 0, handle_default_get },
+	{ GEN_PLVL_OP_RANGE_GET, 0, handle_range_get },
+	BT_MESH_MODEL_OP_END,
+};
+
+const struct bt_mesh_model_op _gen_plvl_setup_srv_op[] = {
+	{ GEN_PLVL_OP_DEFAULT_SET, 2, handle_default_set },
+	{ GEN_PLVL_OP_DEFAULT_SET_UNACK, 2, handle_default_set_unack },
+	{ GEN_PLVL_OP_RANGE_SET, 4, handle_range_set },
+	{ GEN_PLVL_OP_RANGE_SET_UNACK, 4, handle_range_set_unack },
+	BT_MESH_MODEL_OP_END,
+};
+
+static void lvl_get(struct gen_lvl_srv *lvl_srv, struct bt_mesh_msg_ctx *ctx,
+		    struct gen_lvl_status *rsp)
+{
+	struct gen_plvl_srv *srv =
+		CONTAINER_OF(lvl_srv, struct gen_plvl_srv, lvl);
+	struct gen_plvl_status status = { 0 };
+
+	srv->handlers->power_get(srv, ctx, &status);
+
+	rsp->present = POWER_TO_LVL(status.present);
+	rsp->target = POWER_TO_LVL(status.target);
+	rsp->remaining_time = status.remaining_time;
+}
+
+static void lvl_set(struct gen_lvl_srv *lvl_srv, struct bt_mesh_msg_ctx *ctx,
+		    const struct gen_lvl_set *lvl_set,
+		    struct gen_lvl_status *rsp)
+{
+	struct gen_plvl_srv *srv =
+		CONTAINER_OF(lvl_srv, struct gen_plvl_srv, lvl);
+	struct gen_plvl_set set = {
+		.power = LVL_TO_POWER(lvl_set->level),
+		.transition = lvl_set->transition,
+	};
+	struct gen_plvl_status status;
+
+	change_level(srv, ctx, &set, &status);
+
+	if (rsp) {
+		rsp->present = POWER_TO_LVL(status.present);
+		rsp->target = POWER_TO_LVL(status.target);
+		rsp->remaining_time = status.remaining_time;
+	}
+}
+
+static void lvl_delta_set(struct gen_lvl_srv *lvl_srv,
+			  struct bt_mesh_msg_ctx *ctx,
+			  const struct gen_lvl_delta_set *delta_set,
+			  struct gen_lvl_status *rsp)
+{
+	struct gen_plvl_srv *srv =
+		CONTAINER_OF(lvl_srv, struct gen_plvl_srv, lvl);
+	struct gen_plvl_status status = { 0 };
+
+	srv->handlers->power_get(srv, NULL, &status);
+
+	struct gen_plvl_set set = {
+		.power = status.present + delta_set->delta,
+		.transition = delta_set->transition,
+	};
+
+	change_level(srv, ctx, &set, &status);
+
+	if (rsp) {
+		rsp->present = POWER_TO_LVL(status.present);
+		rsp->target = POWER_TO_LVL(status.target);
+		rsp->remaining_time = status.remaining_time;
+	}
+}
+
+static void lvl_move_set(struct gen_lvl_srv *lvl_srv,
+			 struct bt_mesh_msg_ctx *ctx,
+			 const struct gen_lvl_move_set *move_set,
+			 struct gen_lvl_status *rsp)
+{
+	struct gen_plvl_srv *srv =
+		CONTAINER_OF(lvl_srv, struct gen_plvl_srv, lvl);
+	struct gen_plvl_status status = { 0 };
+	u16_t target;
+
+	srv->handlers->power_get(srv, NULL, &status);
+
+	if (move_set->delta > 0) {
+		target = srv->range.max;
+	} else if (move_set->delta < 0) {
+		target = srv->range.min;
+	} else {
+		target = status.present;
+	}
+
+	struct gen_plvl_set set = {
+		.power = target,
+	};
+	struct model_transition transition;
+
+	if (move_set->delta != 0 && move_set->transition) {
+		s32_t delta = target - status.present;
+
+		s32_t time_to_edge =
+			(delta * move_set->transition->time) / move_set->delta;
```

Comment:
This does not look correct. Move speed depends upon the Delta field divided by given transition time. Here, the assumption is made that the power level server is going to stop at max or min edges. That may not be the desired application use case. The application may want to cycle through the power level for an infinite time after receiving a move message (that is how Level move behaviour is). 

I would suggest to provide `power_move_set` callback and pass the power values and transition information to the application, in order to provide implementers with a chance to implement move transition for power level as per their requirement.

Reply:
Actually, the power level overflow behavior is defined in Section 3.1.5.1.1 of the Model spec:

> The Generic Power Actual state shall not wrap around (i.e., from 65535 to 0, or 0 to 65535) when it reaches the maximum or minimum value.

So we can make the assumption that a move is a transition towards max or min, and any additional callbacks should be unnecessary.

Reply:
Ok. Then the assumption is fine. However, the calculation for the speed here seems incorrect. For move message, 
Speed = requested delta/transition time. 

Here, you are doing,
Speed = (Target (i.e min/max/0) - present) / transition time => which will be quite different than what is expected by the user.

Reply:
Should be resolved with the `abs`.

Reply:
resolved.